clients/js/src/signed_calls.ts

```diff
@@ -21,8 +18,68 @@ const DEFAULT_GAS_LIMIT = 30_000_000;
 const DEFAULT_VALUE = 0;
 const DEFAULT_DATA = '0x';
 const zeroAddress = () => `0x${'0'.repeat(40)}`;
+const MAX_SAFE_BIG_NUM = 0x1fffffffffffff - 1;
 
-export type Signer = Pick<EthersSigner, 'getTransactionCount' | 'getChainId'> &
+class SignedCallCacheManager {
+  public enabled = true;
+
+  // for each signer, we cache the signature of the hash of each SignableCall
+  private cachedSignatures = new Map<string, Map<string, Uint8Array>>();
+  // for each ChainId, we cache the base block number to make the same leash
+  private cachedBlockNumbers = new Map<number, BlockTag>();
+
+  public enable() {
+    this.enabled = true;
+  }
+
+  public disable() {
+    this.enabled = false;
+    cachedSignatures.clear();
+    cachedBlockNumbers.clear();
+  }
+
+  public setSignatureCache(
+    address: string,
+    hash: string,
+    signature: Uint8Array,
+  ) {
+    if (!this.enabled) return;
+    if (!this.cachedSignatures.has(address))
+      this.cachedSignatures.set(address, new Map<string, Uint8Array>());
+    this.cachedSignatures.get(address)!.set(hash, signature);
+  }
+
+  public getSignatureCache(
+    address: string,
+    hash: string,
+  ): Uint8Array | undefined {
+    return this.cachedSignatures.get(address)?.get(hash);
+  }
+
+  public setBlockNum(chainId: number, blockNum: BlockTag) {
+    if (!this.enabled) return;
+    this.cachedBlockNumbers.set(chainId, blockNum);
+  }
+
+  public getBlockNum(chainId: number): BlockTag | undefined {
+    return this.cachedBlockNumbers.get(chainId);
+  }
+}
+
+const _cacheManager = new SignedCallCacheManager();
+
+export function enableSignedCallCaches() {
+  _cacheManager.enable();
+}
+
+export function disableSignedCallCaches() {
+  _cacheManager.disable();
+}
```

Comment:
```suggestion
```
Please don't add side effects, as it can confuse bundlers. It's better to initialize in wrap if needed.